### PR TITLE
RADIO Demo for kNN Classification

### PR DESCRIPTION
**Pull request by mranzinger**

### CODE REVIEW

File: examples/knn_classification.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    num_train_steps = _round_up(num_train_examples, args.batch_size * world_size)
+    num_eval_steps = _round_up(num_eval_examples, args.batch_size * world_size)
+
+    def _get_dataset(builder, split: str):
+        dataset = builder.as_dataset(split=split)
+        dataset = dataset.to_iterable_dataset(num_shards=world_size * max(1, args.workers))
+        dataset = split_dataset_by_node(dataset, rank=rank, world_size=world_size)
+        dataset = dataset.map(lambda ex: dict(image=transform(ex['image']), label=torch.as_tensor(ex['label'], dtype=torch.int64)))
+
+        loader = DataLoader(dataset, batch_size=args.batch_size, shuffle=False,
```
Comment (gheinrich) [inline examples/knn_classification.py:134]: Is it legal to use batch_size>1 when we don't do center crops?

File: examples/knn_classification.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+        for s in samples
+    ]
+
+    size_groups = defaultdict(lambda: [[],[]])
+    for im, lab in zip(images, labels):
+        grp = size_groups[im.shape]
+        grp[0].append(im)
+        grp[1].append(lab)
+
+    ret = [
```
Comment (gheinrich) [inline examples/knn_classification.py:256]: So the data loader returns sub-batches of different image shapes? Which allows us to do the inference in one pass for each sub-batch within the batch?

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+$$
+
+Following Wu et.al. we set $\tau = 0.07$, thus $c_{\mathbf{q}}$ is the predicted class. Note that this is the same formulation that [DINO](http://arxiv.org/abs/2104.14294) and [DINOv2](http://arxiv.org/abs/2304.07193) use.
+
+We argue that K-NN classification is a great evaluation protocol for holistic image tasks because it works for any vision model that can produce a summary vector for an image, and thus it easily allows us to compare, for example, CLIP vs DINOv2 vs SAM vs etc. The technique is not zero-shot like CLIP in the sense that it still requires a scan through the training set, however, it's also a training-free technique wrt the model.
+
+### Evaluation Script
+
+We provide an implementation of the K-NN algorithm with with the `knn_classification.py` script, which uses the HuggingFace Datasets library to make evaluation simple. This script works in both single-GPU and multi-GPU setups. For multi-GPU, replace `python` with `torchrun`.
+
```
Comment (gheinrich) [inline examples/README.md:25]: maybe we could add an example multi-GPU invocation to the README?

File: examples/knn_classification.py (hunk 1 above)
Comment (mranzinger) [reply examples/knn_classification.py:134]: The collater groups the images into size bins. So "batch_size" becomes more fluid in the sense that we load that many examples, group them based on size, and then process those parts.

File: examples/knn_classification.py (hunk 2 above)
Comment (mranzinger) [reply examples/knn_classification.py:256]: Yep, exactly

File: examples/README.md (hunk 3 above)
Comment (mranzinger) [reply examples/README.md:25]: I'm a bit worried that it's environment dependent. I'm not sure if the specific way that I'm running it on our cluster is a general setup, or something that we're doing at NVIDIA. The docs for `torchrun` seem to suggest that  it "just works" generally.

Comment (gheinrich) [reply examples/README.md:25]: Yes I find torchrun to be more portable than mpirun :-) - In single node setups, I use `--nproc_per_node` to specify the number of GPUs.

Comment (mranzinger) [reply examples/README.md:25]: Is `--nproc_per_node` necessary? Or if I just launch with no settings, would that pick up all GPUs on the system?

Comment (gheinrich) [reply examples/README.md:25]: I think `--nproc-per-node` is necessary if nothing else sets up the environment variables for distributed execution.

Comment (mranzinger) [reply examples/README.md:25]: So `torchrun --nproc-per-node 8 knn_classification.py ...` is likely a sufficient command?

Comment (gheinrich) [reply examples/README.md:25]: Yes is my go-to incantation :-)

Comment (mranzinger) [reply examples/README.md:25]: Cool. Done

**Review by gheinrich — APPROVED** [review]

**Pull request merged**